### ISSUE TITLE
localText update does not re-render the grid with new translation

### ISSUE THREAD

**Issue by alexfauquette** [mui/mui-x#4119]
### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] I have tested the latest version

### Current behavior 😯

From a [stackoverflow message](https://stackoverflow.com/questions/70782686/mui-datagrid-renders-the-translated-fields-on-change-but-is-not-reflected-in-ui)

https://codesandbox.io/s/datagridprodemo-material-demo-forked-ozkkjv?file=/demo.js

When changing the `localeText` the grid does not update. It displays the previous language. To have the correct one, we have to for a rerender by clicking on the grid

### Expected behavior 🤔

When `localText` prop is updated, the components re-render texts

### Your environment 🌎

<details>
  <summary>`npx @mui/envinfo`</summary>

```
  Don't forget to mention which browser you used.
  Output from `npx @mui/envinfo` goes here.
```
</details>

**Comment by flaviendelangle** [comment]
`apiRef.current.getLocaleText` calls an outdated version of `getLocaleText` which do not yet have the new locals in its scope

That's weird, `useGridApiMethod` is supposed to handle this kind of update correctly

**Comment by flaviendelangle** [comment]
Related to https://github.com/facebook/react/issues/14099#issuecomment-440013892

**Comment by flaviendelangle** [comment]
With React 18, updating a ref in the render is unsafe.
Because you can have tearing (see [this discussion](https://github.com/reactwg/react-18/discussions/69))

And I don't see how to fix this problem in a React 18 compatible way, without deep changes

**Comment by m4theushw** [comment]
I took a look at this issue and I noticed that it could happen with other props as well if a rerender doesn't occur after the prop change. In the first render after changing the locale, `getLocaleText` is called and it returns the English locale. Then, the effect below is called and `apiMethodsRef` is updated. If there was a 2nd render here, `getLocaleText` would return the Hindi locale, but there's none.

https://github.com/mui/mui-x/blob/ffcc4cf69c0f3cd2df760cb1969d00200f1be65c/packages/grid/x-data-grid/src/hooks/utils/useGridApiMethod.ts#L26-L28

This issue doesn't occur with other props only because we usually call `forceUpdate` after syncing the state. To proof this, apply the diff below and access `selectionModel` inside an API method during the render phase. It will be called only once and with the outdated value.

```diff
diff --git a/packages/grid/x-data-grid/src/hooks/features/selection/useGridSelection.ts b/packages/grid/x-data-grid/src/hooks/features/selection/useGridSelection.ts
index 4c71685a..16167707 100644
--- a/packages/grid/x-data-grid/src/hooks/features/selection/useGridSelection.ts
+++ b/packages/grid/x-data-grid/src/hooks/features/selection/useGridSelection.ts
@@ -130,8 +130,6 @@ export const useGridSelection = (
       const currentModel = gridSelectionStateSelector(apiRef.current.state);
       if (currentModel !== model) {
         logger.debug(`Setting selection model`);
-        apiRef.current.setState((state) => ({ ...state, selection: model }));
-        apiRef.current.forceUpdate();
       }
     },
     [apiRef, logger],
```

As a workaround we could trigger a rerender manually once `localeText` changes. Or do IHMO the right thing which is to replace the ref with a state. Then, we ensure that the grid will rerender if `apiMethodsRef` changes. The problem with this last option is that we'll have to go through all hooks to memoize the API methods.

**Comment by mracette** [comment]
This is a bit of a sore spot for us as we just upgraded to Pro. Are there any workarounds that we can use locally while we await a long-term fix?

**Comment by PrateekPadhy** [comment]
We are also facing a similar issue.

https://codesandbox.io/s/focused-artem-fytv9q?file=/demo.tsx

**Comment by Fruitseller** [comment]
We have the same problem.

**Comment by alexfauquette** [comment]
@Fruitseller This issue is more than two years old. To simplify issue traking we do not open old ones.

 Could you open a new issue showing your problem with an updated version of the grid?
